stan/math/rev/fun/generalized_inverse.hpp

```diff
@@ -0,0 +1,124 @@
+#ifndef STAN_MATH_REV_FUN_GENERALIZED_INVERSE_HPP
+#define STAN_MATH_REV_FUN_GENERALIZED_INVERSE_HPP
+
+#include <stan/math/rev/core.hpp>
+#include <stan/math/prim/err.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/rev/fun/cholesky_decompose.hpp>
+#include <stan/math/prim/fun/transpose.hpp>
+#include <stan/math/rev/fun/tcrossprod.hpp>
+#include <stan/math/prim/fun/crossprod.hpp>
+#include <stan/math/prim/fun/inverse_spd.hpp>
+#include <stan/math/rev/fun/mdivide_left_spd.hpp>
+#include <stan/math/prim/fun/mdivide_right_spd.hpp>
+#include <stan/math/rev/fun/inverse.hpp>
+
+namespace stan {
+namespace math {
+
+/**
+ * Reverse mode differentiation algorithm reference:
+ *
+ * The Differentiation of Pseudo-Inverses and Nonlinear Least Squares Problems
+ * Whose Variables Separate. Author(s): G. H. Golub and V. Pereyra. Source: SIAM
+ * Journal on Numerical Analysis, Vol. 10, No. 2 (Apr., 1973), pp. 413-432
+ *
+ * Equation 4.12 in the paper
+ */
+template <typename EigMat, require_rev_matrix_t<EigMat>* = nullptr>
```

Comment:
@SteveBronder , I added `require_rev_matrix_t` here. Let me know if it's the right signature.